pkg/operator/starter.go

```diff
@@ -21,91 +21,137 @@ import (
+	isHypershift := guestKubeConfigString != ""
+	if isHypershift {
+		guestKubeConfig, err = client.GetKubeConfigOrInClusterConfig(guestKubeConfigString, nil)
+		if err != nil {
+			return err
+		}
+		guestKubeClient = kubeclient.NewForConfigOrDie(rest.AddUserAgent(guestKubeConfig, operatorName))
+
+		// Create all events in the GUEST cluster.
+		// Use name of the operator Deployment in the management cluster + namespace
+		// in the guest cluster as the closest approximation of the real involvedObject.
+		controllerRef, err := events.GetControllerReferenceForCurrentPod(ctx, controlPlaneKubeClient, controlPlaneNamespace, nil)
+		controllerRef.Namespace = guestNamespace
+		if err != nil {
+			klog.Warningf("unable to get owner reference (falling back to namespace): %v", err)
+		}
+		eventRecorder = events.NewKubeRecorder(guestKubeClient.CoreV1().Events(guestNamespace), operandName, controllerRef)
+	}
+
+	guestDynamicClient, err := dynamic.NewForConfig(guestKubeConfig)
 	if err != nil {
 		return err
 	}
 
-	csiControllerSet := csicontrollerset.NewCSIControllerSet(
-		operatorClient,
-		controllerConfig.EventRecorder,
+	// Client informers for the GUEST cluster.
+	guestKubeInformersForNamespaces := v1helpers.NewKubeInformersForNamespaces(guestKubeClient, guestNamespace, "")
+	guestConfigMapInformer := guestKubeInformersForNamespaces.InformersFor(guestNamespace).Core().V1().ConfigMaps()
+	guestNodeInformer := guestKubeInformersForNamespaces.InformersFor("").Core().V1().Nodes()
+
+	guestConfigClient := configclient.NewForConfigOrDie(rest.AddUserAgent(guestKubeConfig, operatorName))
+	guestConfigInformers := configinformers.NewSharedInformerFactory(guestConfigClient, 20*time.Minute)
+	guestInfraInformer := guestConfigInformers.Config().V1().Infrastructures()
+
+	// Create client and informers for our ClusterCSIDriver CR.
+	gvr := opv1.SchemeGroupVersion.WithResource("clustercsidrivers")
+	guestOperatorClient, guestDynamicInformers, err := goc.NewClusterScopedOperatorClientWithConfigName(guestKubeConfig, gvr, string(opv1.AWSEBSCSIDriver))
+	if err != nil {
+		return err
+	}
+
+	controlPlaneInformersForEvents := []factory.Informer{
+		controlPlaneSecretInformer.Informer(),
+		controlPlaneConfigMapInformer.Informer(),
+		guestNodeInformer.Informer(),
+		guestInfraInformer.Informer(),
+	}
+	if !isHypershift {
+		controlPlaneInformersForEvents = append(controlPlaneInformersForEvents, controlPlaneCloudConfigInformer.Informer())
+	}
+
+	// Start controllers that manage resources in the MANAGEMENT cluster.
+	controlPlaneCSIControllerSet := csicontrollerset.NewCSIControllerSet(
+		guestOperatorClient,
+		eventRecorder,
 	).WithLogLevelController().WithManagementStateController(
 		operandName,
 		false,
 	).WithStaticResourcesController(
 		"AWSEBSDriverStaticResourcesController",
-		kubeClient,
-		dynamicClient,
-		kubeInformersForNamespaces,
-		assets.ReadFile,
+		controlPlaneKubeClient,
+		controlPlaneDynamicClient,
```

Comment:
The static controller accepts kubeclient of control-plane and hence proceeds to required rbac rules associated with attacher, provisioner etc in mgmt cluster and I don't think that will work.

Reply:
See https://github.com/openshift/aws-ebs-csi-driver-operator/pull/159/files#diff-0d623dfd885adb20f991bda4c2453aebd732ca6dbb4d1d4be6e79805c3b48de6R547 (starter.go, line 547). Sidecars running in the mgmt cluster have admin permissions to the hosted cluster, so those RBAC rules don't matter. We'll need to drop them in Hypershift eventually.

Reply:
The operator should not create kube-rbac ClusterRoles in the mgmt cluster - it does not have permissions to read/create ClusterRoles there at all.

```
 			"rbac/prometheus_role.yaml",
			"rbac/prometheus_rolebinding.yaml",
			"rbac/kube_rbac_proxy_role.yaml",
			"rbac/kube_rbac_proxy_binding.yaml",
```

Reply:
Creating those only on standalone clusters.